### PR TITLE
Beef up infer-mut-hack

### PR DESCRIPTION
**Pull request by ranjitjhala**
(To avoid it getting "confused" be extraneous `Constr`)

### CODE REVIEW

File: crates/flux-refineck/src/checker.rs [hunk 1]
```diff
@@ -1704,6 +1707,29 @@ fn all_predicates_of(
     .flatten()
 }
 
+struct SkipConstr;
```
Comment (nilehmann) [inline crates/flux-refineck/src/checker.rs:1710]: @ranjitjhala 

We already have [Ty::unconstr](https://flux-rs.github.io/flux/doc/flux_middle/rty/struct.Ty.html#method.unconstr) which is a more general version of this.

Comment (ranjitjhala) [reply crates/flux-refineck/src/checker.rs:1710]: Interesting -- didn't know about it! 

But, not really, replacing

```rust
let ty = SkipConstr.fold_ty(ty);
```

with

```rust
 let (ty, _) = ty.unconstr();
```

Doesn't seem to work -- looks like `unconstr` is not "deep": it expects the `Constr` at the very top?

Comment (ranjitjhala) [reply crates/flux-refineck/src/checker.rs:1710]: Looks like `unconstr` just peels away all the "topmost" `Constr` -- the moment it hits something else it stops. 

So it won't work with 

```rust
#[flux::sig(fn (xs: &mut { [i32][@n] | n > 1}))]
fn lib1(_xs: &mut [i32]) {}
```

Comment (nilehmann) [reply crates/flux-refineck/src/checker.rs:1710]: You first need to match on the `&mut T` and then `unconstr` the `T`

```rust
 if let rty::Ref!(_, deref_ty, Mutability::Mut) = ty.kind()
        && let (deref_ty, _) = deref_ty.unconstr()
        && let TyKind::Indexed(..) = inner_ty.kind()
```

I wonder though if there's a clean way to handle this during [_canonicalization_](https://github.com/flux-rs/flux/blob/main/crates/flux-middle/src/rty/canonicalize.rs) (the thing `unpack` is implementing on top of). It is always safe to turn an `&mut {T | p}` into `{&mut T | p}` so we could get rid of it during `unpack`

Comment (ranjitjhala) [reply crates/flux-refineck/src/checker.rs:1710]: This is why I have two tests -- the other one flips the order which fails your method

```rust
#[flux::sig(fn (xs: &mut { [i32][@n] | n > 1}))]
fn lib1(_xs: &mut [i32]) {}

#[flux::sig(fn (xs: { &mut  [i32][@n] | n > 1}))]
fn lib2(_xs: &mut [i32]) {}
```

But yes, if you had `canonicalized` presumably `lib1` gets turned into `lib2`?

Comment (nilehmann) [reply crates/flux-refineck/src/checker.rs:1710]: ah yeah, you'd need to unconstr twice

```rust
let (ty, _) = ty.unconstr();
if let rty::Ref!(_, deref_ty, Mutability::Mut) = ty.kind() 
    && let (deref_ty, _) = deref_ty.unconstr() 
    && let TyKind::Indexed(..) = inner_ty.kind()
```

I still think there has to be a nice way in which we can canonicalize types before pattern matching on them. The canonicalize infra is almost there...

**Pull request merged**